File: native/chu/tasoller_p.go
package chu

import "math/bits"

type TasollerPlusParser struct{}

func (p *TasollerPlusParser) GetName() string {
	return "TasollerPlus"
}

func (p *TasollerPlusParser) Parse(raw []byte) ChusanData {
	if len(raw) < 36 {
		return ChusanData{}
	}

	data := ChusanData{}
	irButtons := raw[3]
	reversed := bits.Reverse8(irButtons)

	data.Test = (reversed & 1) == 1
	data.Service = (reversed&2)>>1 == 1

	beams := reversed >> 2
	for i := 0; i < 6; i++ {
		data.Air[i] = (beams >> i) & 1 == 1
	}

	copy(data.Slider[:], raw[4:36])

	return data
}

File: native/src/chu/tasoller_p.rs
use super::{ChusanData, ChusanParser};

pub struct TasollerPlusParser;

impl ChusanParser for TasollerPlusParser {
    fn get_name(&self) -> &'static str {
        "TasollerPlus"
    }

    fn parse(&self, raw: &[u8]) -> ChusanData {
        if raw.len() < 36 {
            return ChusanData::default();
        }

        let mut data = ChusanData::default();

        let ir_buttons = raw[3];
        let reversed = ir_buttons.reverse_bits();

        data.test = (reversed & 1) == 1;
        data.service = ((reversed & 2) >> 1) == 1;

        let beams = reversed >> 2;
        for i in 0..6 {
            data.air[i] = ((beams >> i) & 1) == 1;
        }

        data.slider.copy_from_slice(&raw[4..36]);

        data
    }
}
